src/bloqade/pyqrack/squin/runtime.py

```diff
@@ -0,0 +1,323 @@
+from typing import Any
+from dataclasses import dataclass
+
+import numpy as np
+from kirin.dialects import ilist
+
+from pyqrack.pauli import Pauli
+from bloqade.pyqrack import PyQrackQubit
+
+
+@dataclass(frozen=True)
+class OperatorRuntimeABC:
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        raise NotImplementedError(
+            "Operator runtime base class should not be called directly, override the method"
+        )
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        raise RuntimeError(f"Can't apply controlled version of {self}")
+
+    def broadcast_apply(self, qubits: ilist.IList[PyQrackQubit, Any], **kwargs) -> None:
+        for qbit in qubits:
+            if not qbit.is_active():
+                continue
+
+            self.apply(qbit, **kwargs)
+
+
+@dataclass(frozen=True)
+class NonBroadcastableOperatorRuntimeABC(OperatorRuntimeABC):
+    def broadcast_apply(self, qubits: ilist.IList[PyQrackQubit, Any], **kwargs) -> None:
+        raise RuntimeError(
+            f"Operator of type {type(self).__name__} is not broadcastable!"
+        )
+
+
+@dataclass(frozen=True)
+class OperatorRuntime(OperatorRuntimeABC):
+    method_name: str
+
+    def get_method_name(self, adjoint: bool, control: bool) -> str:
+        method_name = ""
+        if control:
+            method_name += "mc"
+
+        if adjoint and self.method_name in ("s", "t"):
+            method_name += "adj"
+
+        return method_name + self.method_name
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        method_name = self.get_method_name(adjoint=adjoint, control=False)
+        getattr(qubits[0].sim_reg, method_name)(qubits[0].addr)
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        ctrls = [qbit.addr for qbit in qubits[:-1]]
+        target = qubits[-1]
+        method_name = self.get_method_name(adjoint=adjoint, control=True)
+        getattr(target.sim_reg, method_name)(ctrls, target.addr)
```

Comment:
```suggestion
        ctrls = [qbit.addr for qbit in qubits[:-1], if qbit.is_active()]
        if len(ctrls) == 0: # do not apply if there are no control qubits. 
            return
        target = qubits[-1]
        method_name = self.get_method_name(adjoint=adjoint, control=True)
        getattr(target.sim_reg, method_name)(ctrls, target.addr)
```

Reply:
@weinbe58 this still applies to the target even if `target.is_active() is False`. That's inconsistent with the `apply` method now though, right?

Reply:
If it's a Rydberg-interaction-type gate, then a missing atom acts like the |0> state.  Often this is equivalent to no gate occurring.